### PR TITLE
Allow specifying custom versioning schemes

### PR DESCRIPTION
**Pull request by azdrojowa123**
Fixes https://github.com/VirtusLab/bazel-steward/issues/33

### CODE REVIEW

File: common/src/main/kotlin/org/virtuslab/bazelsteward/common/UpdateLogic.kt [hunk 1]
```diff
@@ -30,4 +35,21 @@ class UpdateLogic {
...
+
+  private fun <Lib : LibraryId> getVersioningForLibrary(library: Library<Lib>): VersioningSchema {
+    return when (val libraryId = library.id) {
+      is MavenLibraryId -> {
+        val matchingDependencies = bazelStewardConfig.maven.ruledDependencies.filter { it.id == libraryId }
+        when {
+          matchingDependencies.isEmpty() -> VersioningSchema(VersioningType.LOOSE.name)
+          matchingDependencies.size == 1 -> matchingDependencies.first().versioning
+          else -> throw Exception(
+            "In the configuration file more than one ruled dependency is declared for " +
```
Comment (lukaszwawrzyk) [inline common/src/main/kotlin/org/virtuslab/bazelsteward/common/UpdateLogic.kt:47]: we could just pick first matching, if some other rule would catch this dep.

File: config/src/main/kotlin/org/virtuslab/bazelsteward/config/BazelStewardConfigExtractor.kt [hunk 2]
```diff
@@ -25,34 +31,43 @@ data class MavenConfig(
 
 data class MavenDependency(
   val id: MavenLibraryId,
-  val versioning: String
+  var versioning: VersioningSchema
```
Comment (lukaszwawrzyk) [inline config/src/main/kotlin/org/virtuslab/bazelsteward/config/BazelStewardConfigExtractor.kt:34]: var?

File: core/src/main/kotlin/org/virtuslab/bazelsteward/core/library/SemanticVersion.kt [hunk 3]
```diff
@@ -84,24 +84,31 @@ data class SemanticVersion(
...
-        val values = matchResult.groups as MatchNamedGroupCollection
-        SemanticVersion(
-          values["major"]?.value?.toIntOrNull() ?: 0,
-          values["minor"]?.value?.toIntOrNull() ?: 0,
-          values["patch"]?.value?.toIntOrNull() ?: 0,
-          values["preRelease"]?.value ?: "",
-          values["buildMetaData"]?.value ?: "",
-        )
+    fun fromString(value: String, versioningScheme: VersioningSchema): Option<SemanticVersion> {
+      fun matchToSemanticVersion(regex: Regex): Option<SemanticVersion> {
```
Comment (lukaszwawrzyk) [inline core/src/main/kotlin/org/virtuslab/bazelsteward/core/library/SemanticVersion.kt:93]: make sure that you rebase, we don't use Option, just nulls

File: common/src/main/kotlin/org/virtuslab/bazelsteward/common/UpdateLogic.kt (hunk 1 above)
Comment (azdrojowa123) [reply common/src/main/kotlin/org/virtuslab/bazelsteward/common/UpdateLogic.kt:47]: and inform user about double declaration ?

**Comment by lukaszwawrzyk** [comment]
Fixes #33

**Review by lukaszwawrzyk — APPROVED** [review]

**Pull request merged**